### ISSUE TITLE
How to purge cache selectively?

### ISSUE THREAD

**Issue by vspinu** [rstudio/renv#255]
I have ended up (not exactly sure how) in the following situation:

```
renv::status()
The following packages have incorrect hashes:

	askpass        1.1       [Hash: 3af2ffe5d380157819726a25b94cdb0a != 46678e19f261db289be40292d8e5d5bf]
	assertthat     0.2.1     [Hash: 931d1f871e6f91598b94ea7da81cdb98 != 19fb3d0ad048b4eb1a05cacffbae1f66]
	backports      1.1.4     [Hash: e8a88fdd2aa7bea4024f3d7005bb36f4 != 2a78bf0825e0088e117520553ceb7fa5]
	base64enc      0.1-3     [Hash: 9269738ddbafc841f69df4fe9e88df13 != d355963b4b1c3039bfe8d65611a11cca]
	BBmisc         1.11      [Hash: 9a4102c0a258c514bdc1be6fd64f22aa != 34b4d5e7771f029012a2d1ffe496be84]
	BH             1.69.0-1  [Hash: ec8b4d89dd49ae708c4f73d951e8d58e != 0fde015f5153e51df44981da0767f522]
	bitops         1.0-6     [Hash: 2c3ed09bc56cc74a15a5551a454d4963 != 62408262624d9f8e436f51ca901424e2]
	broom          0.5.2     [Hash: 986a4911418f130eb7b88157a5ff4301 != a6cfb50e5172a20815db00739aab4e78]
	callr          3.3.1     [Hash: 9586ca46431b6a33da4afd1383907528 != b9788f81c78e3187d199c4057872fe06]
	caTools        1.17.1.2  [Hash: 11635e904c672a1a3e04fd1f67b42ec0 != 54626244d6196ab491e26d5b9590dff6]
	cellranger     1.1.0     [Hash: 059f86e434cc472a8443f7d5766a89c8 != 2cd2a065192f3ba08a919d8800dba7e0]
	checkmate      1.9.4     [Hash: 949fc96697233fec1e11ceb7c58ff499 != 51788d50ba2a6ac3bf71efadaf1f76d9]
	cli            1.1.0     [Hash: 8def44b1aa7cfb998b3bdd0fbfed6226 != 9b3dc7798f53ba3050b41ce4a3febdc5]
	clipr          0.7.0     [Hash: 468a2220579e9889dd831b8badf87226 != 9a99b54d2baf9393d0a68b6ea9513ee6]

... many more ...
```

What does this mean and what is the recommended way to go about this? 

Thanks!

**Comment by kevinushey** [comment]
For context, when `renv` places a package into the package cache, it creates a path based on a hash (MD5 sum) of some contents of the `DESCRIPTION` file. For example:

```
> pkgpaths <- renv:::renv_cache_list(packages = "digest")
> print(pkgpaths)
[1] "/Users/kevinushey/Library/Application Support/renv/cache/v4/R-3.5/x86_64-apple-darwin15.6.0/digest/0.6.20/ee526dc5fbf649c15ce452a474543a4c/digest"
[2] "/Users/kevinushey/Library/Application Support/renv/cache/v4/R-3.5/x86_64-apple-darwin15.6.0/digest/0.6.21/80c4e8b2c804c5ffdd23f2f9d5765161/digest"
[3] "/Users/kevinushey/Library/Application Support/renv/cache/v4/R-3.5/x86_64-apple-darwin15.6.0/digest/0.6.22/1417fd7efe95eb5859e69d1cf3ccaaaa/digest"
```

Normally, these hashes should not change post-install. For example:

```
> descpaths <- file.path(pkgpaths, "DESCRIPTION")
> hashes <- renv:::map_chr(descpaths, renv:::renv_hash_description)
> names(hashes) <- substring(names(hashes), nchar(renv:::renv_paths_cache()) + 1)
> hashes
/digest/0.6.20/ee526dc5fbf649c15ce452a474543a4c/digest/DESCRIPTION 
                                "ee526dc5fbf649c15ce452a474543a4c" 
/digest/0.6.21/80c4e8b2c804c5ffdd23f2f9d5765161/digest/DESCRIPTION 
                                "80c4e8b2c804c5ffdd23f2f9d5765161" 
/digest/0.6.22/1417fd7efe95eb5859e69d1cf3ccaaaa/digest/DESCRIPTION 
                                "1417fd7efe95eb5859e69d1cf3ccaaaa" 
```

This would likely imply that the DESCRIPTION files for these packages may have been modified after they were installed into the cache.

Since by default, packages used within a project are just symlinks into the cache, if the installed packages were modified post-install and post-cache, you might see something like this.

If at all possible, it would be helpful to know how you got into this state in case it's a bug in `renv` or otherwise an issue that `renv` could better handle.

Right now, `renv` doesn't provide any tools to re-hash the package cache (so that these paths could be updated) but that seems worth implementing.

**Comment by kevinushey** [comment]
There is the `renv::purge()` function, but it may not be fine-grained enough for your use.

Please also let me know if you're able to discover how `renv` found itself in this state.

**Comment by vspinu** [comment]
I don't recall exactly what I did but it's either:

  - install a bunch of packages from an older MRAN and then re-install with current CRAN
  - change manually version of R in renv.lock and update packages

**Comment by kevinushey** [comment]
It looks like one package in my `renv` cache is in this state as well:

```
> renv::status()
The following packages have incorrect hashes:

        curl 4.1 [Hash: cdd60bbfb548290b3a6f02f8f4cd9312 != 752c16159f6bc8c3f90ac9fb067bde4e]

These packages should be purged and re-installed.
```

It looks like this is caused by some kind of cross-talk by `renv` with different versions of R:

```
# R 3.5
$ R --slave -e 'renv:::renv_hash_description(file.path(getwd(), "curl_4.1.tar.gz"))'
[1] "752c16159f6bc8c3f90ac9fb067bde4e"

# R 3.6
$ R --slave -e 'renv:::renv_hash_description(file.path(getwd(), "curl_4.1.tar.gz"))'
[1] "cdd60bbfb548290b3a6f02f8f4cd9312"
```

I'll see if I can figure out why this has occurred.

**Comment by vspinu** [comment]
So it looks like it's the version of R that was causing it. We can close this for now and if the issue re-occurs I will be back with you.

**Comment by vspinu** [comment]
I have got into this state again. On a large scale. This time not because of different versions of R, but probably because of different versions of R env. I upgraded renv multiple times and merged different renv.lock versions from a colaborator two times with conflicting versions.

**Comment by kevinushey** [comment]
Sorry for the trouble.

The hash format has indeed changed now; previously, the active R version was included as part of the hash generated for a particular package, which in hindsight was not the right decision. This change does unfortunately necessitate re-caching of packages.

If you've updated to the latest version of `renv`, you can use:

```
renv::rehash()
```

to re-cache any cached packages. Note that we leave behind symlinks for the old package paths so that projects using the cache with older versions of `renv` can still function.

**Comment by vspinu** [comment]
I see. Funny part is that once I have upgraded renv I didn't even had to rehash. The warning disappeared.

**Comment by vspinu** [comment]
For the record I got this again after  1) switch to a branch with renv 0.8.3-66 + renv::restore 2) switch to a branch 0.8.3-46 + renv::restore 3) switch to 0.8.3-66 + renv::restore 4) renv::rehash.

The `rehash` doesn't help :(

**Comment by kevinushey** [comment]
Are you sure that you're seeing the warnings + errors when running the latest development version of `renv`?

Any chance you can distill what you're seeing down to a reproducible example?

**Comment by vspinu** [comment]
It would be rather hard. I moved a lot between different version of rlock.lock today and things just pop with no clear pattern. The inability to track things down is simply because they emerge on explicit `renv::status`, and not when they occur, but I am not sure you can improve on this.

I could be faster if I try to track the issue myself locally.

**Comment by vspinu** [comment]
I keep seeing packages being re-installed again and again even though I have surely installed those versions before (multiple times). I think these issues are related.

**Comment by kevinushey** [comment]
At least for myself, it's challenging to fix these issues without a reproducible example -- in particular, it's not clear to me how one could end up in these various states.

**Comment by vspinu** [comment]
Ok, here is what I have. It turned out that all of these hash mismatches come from the symlinks which you have created recently when dropping R version from the hash (if I recall correctly).

This is my typical dir structure in v5 cache directory:

```
.
├── 0.3-2
│   ├── c6d8a035a88c6e3222f0efaac18406f6
│   │   └── whisker -> /home/vspinu/.local/share/renv/cache/v5/R-3.6/x86_64-pc-linux-gnu/whisker/0.3-2/c944abf3f12a97b8369a6f6ba8186d23/whisker
│   └── c944abf3f12a97b8369a6f6ba8186d23
│       └── whisker
└── 0.4
    └── ca970b96d894e90397ed20637a0c1bbe
        └── whisker

```
But `renv_hash_description_impl` ignores this fact and computes the hash from the description, right?

**Comment by kevinushey** [comment]
> But renv_hash_description_impl ignores this fact and computes the hash from the description, right?

That's right. `c6d8a035a88c6e3222f0efaac18406f6` is the hash that's generated with the CRAN release of `renv`, using R 3.6.1;  `c944abf3f12a97b8369a6f6ba8186d23` would be generated by the development version of `renv`.

In other words, it appears as though an older version of `renv` had cached this package, and then `renv::rehash()` was called and migrated the package to the new location (leaving behind a symlink for older `renv` packages which might still require it).

What's not clear to me is how this copy of whisker `c6d8a035a88c6e3222f0efaac18406f6` got into the v5 cache in the first place, since older versions of `renv` should use the v4 cache, and the cache version bump happened only after the cache format was changed.

**Comment by vspinu** [comment]
This I wouldn't know, but I surely called `rehash` on many occasions so it was always called from new version of `renv`, so the bug must lie there. On the other hand, this part is not very important, once people transitioned to new version the problem is gone (I hope). All people have to do is to clear the cache entirely, which might be frustrating to figure out.

**Comment by kevinushey** [comment]
I'm fairly sure this should be resolved now in the development version of `renv` with the v5 hash, but please re-open if you bump into this again and can provide any other context for reproducing the issue.

**Comment by vspinu** [comment]
This problem is still there with ‘0.9.3’. `renv::status` give a bunch of incorrect hashes, but `renv::rehash()` gives me `Your cache is already up-to-date -- nothing to do.`. 

I have removed the v4 cache from my cache repository so it's a problem with V5. 

All those wrong hashes come from older versions of the packages. For instance the problem is shown for R6 v2.4.0, but the renv.lock contains v2.4.1. So my hunch is that `renv::rehash()` only rehashes packages from renv.lock, not all the packages in cache. 

Would be nice to have an enhanced version of `purge` to get rid of old packages btw, but I guess that's a topic for a new issue.

**Comment by kevinushey** [comment]
This is (partly) fixed in the latest builds of `renv`. `renv::rehash()` correctly copies packages to their new, correct location in the cache, but leaves behind the old, incorrect copy. Unfortunately, `renv::status()` in 0.9.3 fails to detect this state and so continues to warn about the old packages.

I decided not to remove the old cache entries just in case any existing project still depended on those packages.

I still need to think about adding better cache management tools to `renv` as well.
